avro2py/codegen.py

```diff
@@ -0,0 +1,443 @@
+""" Avro codegen for Python 3. """
+
+import ast
+import itertools
+import re
+from collections import defaultdict
+from contextlib import contextmanager
+from textwrap import wrap
+from typing import NamedTuple, List, Union, Generator, Tuple, Optional, Dict
+
+from avro2py.avro_types import (
+    Primitives, LogicalTypes, Record, AvroPrimitives, DefinedType, Enum, Array,
+    Map, Fixed, LogicalType,
+)
+
+PRIMITIVE_TYPE_MAPPINGS = {
+    Primitives.NULL: 'None',
+    Primitives.BOOLEAN: 'bool',
+    Primitives.INT: 'int',
+    Primitives.LONG: 'int',  # *
+    Primitives.FLOAT: 'float',
+    Primitives.DOUBLE: 'float',  # *
+    Primitives.BYTES: 'bytes',
+    Primitives.STRING: 'str',
+}
+
+LOGICAL_TYPE_MAPPINGS = {
+    LogicalTypes.DECIMAL: ['decimal', 'decimal.Decimal'],
+    LogicalTypes.DATE: ['datetime', 'datetime.date'],
+    LogicalTypes.TIME_MILLIS: ['datetime', 'datetime.time'],
+    LogicalTypes.TIME_MICROS: ['datetime', 'datetime.time'],
+    LogicalTypes.TIMESTAMP_MILLIS: ['datetime', 'datetime.datetime'],
+    LogicalTypes.TIMESTAMP_MICROS: ['datetime', 'datetime.datetime'],
+    LogicalTypes.DURATION: ['datetime', 'datetime.timedelta'],
+    LogicalTypes.UUID: ['uuid', 'uuid.UUID']
+}
+
+
+class ResolvedClassResult(NamedTuple):
+    resolved_class: ast.ClassDef
+    imports: List[Union[ast.Import, ast.ImportFrom]]
+    new_frontier: List[Union[Record, Enum]]
+
+
+def consolidate_imports(imports: List[Union[ast.Import, ast.ImportFrom]]) -> List[ast.Expr]:
+    """Deduplicate and combine imports, leaving smallest possible set."""
+    deduped_imports = [
+        node
+        for repr_, node in sorted({
+            ast.dump(node): node
+            for node in imports
+        }.items())
+    ]
+
+    def group_key(node):
+        module = getattr(node, 'module', None)
+        level = getattr(node, 'level', None)
+        return module, level
+
+    consolidated_imports = []
+    for (module, level), group in itertools.groupby(deduped_imports, key=group_key):
```

Comment:
does the sort on L49 actually match this group key? i thought `groupby` needed it already sorted

Reply:
You're correct about the `groupby` behavior; the key used for sorting in the above happens to put them in the correct order. For instance, 
```python
from foo import Bar
import baz
from foo import Bat
```
ends up `ast.dump`ing like:
```python
ImportFrom(module='foo', names=[alias(name='Bar', asname=None)], level=0)
Import(names=[alias(name='baz', asname=None)])
ImportFrom(module='foo', names=[alias(name='Bat', asname=None)], level=0)
```
. I also implicitly am relying on module names being unique relative to each file in question; e.g.
```python
from foo import Bar
```
and 
```python
from ..package.foo import Baz
```
render:
```python
ImportFrom(module='foo', names=[alias(name='Bar', asname=None)], level=0)
ImportFrom(module='package.foo', names=[alias(name='Baz', asname=None)], level=2)
```